File: bridge-server/internal/bridges/discord.go
package bridges

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/pocketping/bridge-server/internal/config"
	"github.com/pocketping/bridge-server/internal/types"
)

const discordAPIBase = "https://discord.com/api/v10"

// DiscordBridge sends notifications to Discord
type DiscordBridge struct {
	*BaseBridge
	botToken   string
	channelID  string
	webhookURL string
	username   string
	avatarURL  string
	client     *http.Client
}

// NewDiscordBridge creates a new Discord bridge
func NewDiscordBridge(cfg *config.DiscordConfig) *DiscordBridge {
	return &DiscordBridge{
		BaseBridge: NewBaseBridge("discord"),
		botToken:   cfg.BotToken,
		channelID:  cfg.ChannelID,
		webhookURL: cfg.WebhookURL,
		username:   cfg.Username,
		avatarURL:  cfg.AvatarURL,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// isBotMode returns true if using bot mode (vs webhook mode)
func (b *DiscordBridge) isBotMode() bool {
	return b.botToken != "" && b.channelID != ""
}

// discordEmbed represents a Discord embed
type discordEmbed struct {
	Title       string                `json:"title,omitempty"`
	Description string                `json:"description,omitempty"`
	Color       int                   `json:"color,omitempty"`
	Fields      []discordEmbedField   `json:"fields,omitempty"`
	Footer      *discordEmbedFooter   `json:"footer,omitempty"`
	Timestamp   string                `json:"timestamp,omitempty"`
}

type discordEmbedField struct {
	Name   string `json:"name"`
	Value  string `json:"value"`
	Inline bool   `json:"inline,omitempty"`
}

type discordEmbedFooter struct {
	Text string `json:"text"`
}

// discordMessageResponse is the response from sending a message
type discordMessageResponse struct {
	ID string `json:"id"`
}

// sendMessage sends a message to Discord
func (b *DiscordBridge) sendMessage(content string, embeds []discordEmbed, replyToMessageID string) (string, error) {
	data := map[string]interface{}{}

	if content != "" {
		data["content"] = content
	}
	if len(embeds) > 0 {
		data["embeds"] = embeds
	}
	if b.username != "" {
		data["username"] = b.username
	}
	if b.avatarURL != "" {
		data["avatar_url"] = b.avatarURL
	}
	if replyToMessageID != "" {
		data["message_reference"] = map[string]string{
			"message_id": replyToMessageID,
		}
	}

	var url string
	var headers map[string]string

	if b.isBotMode() {
		url = fmt.Sprintf("%s/channels/%s/messages", discordAPIBase, b.channelID)
		headers = map[string]string{
			"Authorization": fmt.Sprintf("Bot %s", b.botToken),
		}
	} else {
		url = b.webhookURL + "?wait=true"
		headers = map[string]string{}
	}

	body, err := json.Marshal(data)
	if err != nil {
		return "", err
	}

	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := b.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		respBody, _ := io.ReadAll(resp.Body)
		log.Printf("[DiscordBridge] API error %d: %s", resp.StatusCode, string(respBody))
		return "", fmt.Errorf("discord API error: %d", resp.StatusCode)
	}

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var msgResp discordMessageResponse
	if err := json.Unmarshal(respBody, &msgResp); err != nil {
		return "", nil // Webhooks may not return ID
	}

	return msgResp.ID, nil
}

// OnNewSession announces a new chat session
func (b *DiscordBridge) OnNewSession(session *types.Session) error {
	visitorName := session.VisitorID
	if session.Identity != nil && session.Identity.Name != "" {
		visitorName = session.Identity.Name
	}

	embed := discordEmbed{
		Title:       "New Chat Session",
		Description: "A new visitor has started a chat",
		Color:       0x00D4AA, // PocketPing teal
		Fields: []discordEmbedField{
			{Name: "Visitor", Value: visitorName, Inline: true},
		},
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}

	if session.Metadata != nil && session.Metadata.URL != "" {
		embed.Fields = append(embed.Fields, discordEmbedField{
			Name:   "Page",
			Value:  session.Metadata.URL,
			Inline: false,
		})
	}

	_, err := b.sendMessage("", []discordEmbed{embed}, "")
	return err
}

// OnVisitorMessage sends a visitor message to Discord
func (b *DiscordBridge) OnVisitorMessage(message *types.Message, session *types.Session, reply *ReplyContext) (*types.BridgeMessageIDs, error) {
	visitorName := session.VisitorID
	if session.Identity != nil && session.Identity.Name != "" {
		visitorName = session.Identity.Name
	}

	content := fmt.Sprintf("**%s**: %s", visitorName, message.Content)

	if len(message.Attachments) > 0 {
		content += fmt.Sprintf(" _(+%d attachment(s))_", len(message.Attachments))
	}

	replyToMessageID := ""
	if reply != nil && reply.BridgeIDs != nil {
		replyToMessageID = reply.BridgeIDs.DiscordMessageID
	}

	msgID, err := b.sendMessage(content, nil, replyToMessageID)
	if err != nil {
		return nil, err
	}

	if msgID != "" {
		return &types.BridgeMessageIDs{DiscordMessageID: msgID}, nil
	}
	return nil, nil
}

// OnOperatorMessage relays an operator message from another bridge
func (b *DiscordBridge) OnOperatorMessage(message *types.Message, session *types.Session, sourceBridge, operatorName string) error {
	if sourceBridge == "discord" {
		return nil
	}

	name := operatorName
	if name == "" {
		name = "Operator"
	}

	content := fmt.Sprintf("**%s** (via %s): %s", name, sourceBridge, message.Content)
	_, err := b.sendMessage(content, nil, "")
	return err
}

// OnTyping sends a typing indicator
func (b *DiscordBridge) OnTyping(sessionID string, isTyping bool) error {
	if !isTyping || !b.isBotMode() {
		return nil
	}

	url := fmt.Sprintf("%s/channels/%s/typing", discordAPIBase, b.channelID)

	req, err := http.NewRequest("POST", url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", fmt.Sprintf("Bot %s", b.botToken))

	resp, err := b.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	return nil
}

// OnMessageRead handles read receipts (no-op for Discord)
func (b *DiscordBridge) OnMessageRead(sessionID string, messageIDs []string, status types.MessageStatus) error {
	return nil
}

// OnCustomEvent sends a custom event notification
func (b *DiscordBridge) OnCustomEvent(event *types.CustomEvent, session *types.Session) error {
	embed := discordEmbed{
		Title:     fmt.Sprintf("Event: %s", event.Name),
		Color:     0x5865F2, // Discord blurple
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}

	if event.Data != nil {
		data, _ := json.MarshalIndent(event.Data, "", "  ")
		embed.Description = fmt.Sprintf("```json\n%s\n```", string(data))
	}

	_, err := b.sendMessage("", []discordEmbed{embed}, "")
	return err
}

// OnIdentityUpdate sends an identity update notification
func (b *DiscordBridge) OnIdentityUpdate(session *types.Session) error {
	if session.Identity == nil {
		return nil
	}

	embed := discordEmbed{
		Title: "User Identified",
		Color: 0x57F287, // Green
		Fields: []discordEmbedField{
			{Name: "User ID", Value: session.Identity.ID, Inline: true},
		},
	}

	if session.Identity.Name != "" {
		embed.Fields = append(embed.Fields, discordEmbedField{
			Name:   "Name",
			Value:  session.Identity.Name,
			Inline: true,
		})
	}
	if session.Identity.Email != "" {
		embed.Fields = append(embed.Fields, discordEmbedField{
			Name:   "Email",
			Value:  session.Identity.Email,
			Inline: true,
		})
	}

	_, err := b.sendMessage("", []discordEmbed{embed}, "")
	return err
}

// OnAITakeover sends an AI takeover notification
func (b *DiscordBridge) OnAITakeover(session *types.Session, reason string) error {
	embed := discordEmbed{
		Title:       "AI Takeover",
		Description: reason,
		Color:       0xFEE75C, // Yellow
	}

	_, err := b.sendMessage("", []discordEmbed{embed}, "")
	return err
}

// OnVisitorMessageEdited syncs a message edit to Discord (bot mode only)
func (b *DiscordBridge) OnVisitorMessageEdited(sessionID, messageID, content string, bridgeIDs *types.BridgeMessageIDs) (*types.BridgeMessageIDs, error) {
	if !b.isBotMode() || bridgeIDs == nil || bridgeIDs.DiscordMessageID == "" {
		return nil, nil
	}

	url := fmt.Sprintf("%s/channels/%s/messages/%s", discordAPIBase, b.channelID, bridgeIDs.DiscordMessageID)

	data := map[string]interface{}{
		"content": fmt.Sprintf("_(edited)_ %s", content),
	}

	body, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest("PATCH", url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bot %s", b.botToken))

	resp, err := b.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		log.Printf("[DiscordBridge] Edit failed: %d", resp.StatusCode)
		return nil, nil
	}

	return &types.BridgeMessageIDs{DiscordMessageID: bridgeIDs.DiscordMessageID}, nil
}

// OnVisitorMessageDeleted syncs a message delete to Discord (bot mode only)
func (b *DiscordBridge) OnVisitorMessageDeleted(sessionID, messageID string, bridgeIDs *types.BridgeMessageIDs) error {
	if !b.isBotMode() || bridgeIDs == nil || bridgeIDs.DiscordMessageID == "" {
		return nil
	}

	url := fmt.Sprintf("%s/channels/%s/messages/%s", discordAPIBase, b.channelID, bridgeIDs.DiscordMessageID)

	req, err := http.NewRequest("DELETE", url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", fmt.Sprintf("Bot %s", b.botToken))

	resp, err := b.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		log.Printf("[DiscordBridge] Delete failed: %d", resp.StatusCode)
	}

	return nil
}

File: packages/sdk-node/src/bridges/discord.ts
import type { Bridge, BridgeMessageResult } from './types';
import type { PocketPing } from '../pocketping';
import type { Session, Message } from '../types';

/**
 * Discord API response types
 */
interface DiscordMessageResponse {
  id: string;
  [key: string]: unknown;
}

/**
 * Options for Discord webhook mode
 */
export interface DiscordWebhookOptions {
  /** Custom username for webhook messages */
  username?: string;
  /** Custom avatar URL for webhook messages */
  avatarUrl?: string;
}

/**
 * Options for Discord bot mode
 */
export interface DiscordBotOptions {
  /** Custom username displayed in embeds */
  username?: string;
  /** Custom avatar URL for embeds */
  avatarUrl?: string;
}

/**
 * Discord Bridge for PocketPing.
 * Sends visitor messages to a Discord channel using webhooks or bot API.
 *
 * @example Webhook mode
 * ```ts
 * const discord = DiscordBridge.webhook(
 *   'https://discord.com/api/webhooks/123/abc',
 *   { username: 'PocketPing' }
 * );
 * ```
 *
 * @example Bot mode
 * ```ts
 * const discord = DiscordBridge.bot(
 *   'BOT_TOKEN',
 *   'CHANNEL_ID',
 *   { username: 'PocketPing' }
 * );
 * ```
 */
export class DiscordBridge implements Bridge {
  readonly name = 'discord';

  private pocketping?: PocketPing;
  private readonly mode: 'webhook' | 'bot';
  private readonly webhookUrl?: string;
  private readonly botToken?: string;
  private readonly channelId?: string;
  private readonly username?: string;
  private readonly avatarUrl?: string;

  private constructor(config: {
    mode: 'webhook' | 'bot';
    webhookUrl?: string;
    botToken?: string;
    channelId?: string;
    username?: string;
    avatarUrl?: string;
  }) {
    this.mode = config.mode;
    this.webhookUrl = config.webhookUrl;
    this.botToken = config.botToken;
    this.channelId = config.channelId;
    this.username = config.username;
    this.avatarUrl = config.avatarUrl;
  }

  /**
   * Create a Discord bridge using a webhook URL
   */
  static webhook(
    webhookUrl: string,
    options: DiscordWebhookOptions = {}
  ): DiscordBridge {
    return new DiscordBridge({
      mode: 'webhook',
      webhookUrl,
      username: options.username,
      avatarUrl: options.avatarUrl,
    });
  }

  /**
   * Create a Discord bridge using a bot token
   */
  static bot(
    botToken: string,
    channelId: string,
    options: DiscordBotOptions = {}
  ): DiscordBridge {
    return new DiscordBridge({
      mode: 'bot',
      botToken,
      channelId,
      username: options.username,
      avatarUrl: options.avatarUrl,
    });
  }

  /**
   * Initialize the bridge (optional setup)
   */
  async init(pocketping: PocketPing): Promise<void> {
    this.pocketping = pocketping;
    // Verify connection based on mode
    if (this.mode === 'bot' && this.botToken) {
      try {
        const response = await fetch('https://discord.com/api/v10/users/@me', {
          headers: { Authorization: `Bot ${this.botToken}` },
        });
        if (!response.ok) {
          console.error('[DiscordBridge] Invalid bot token');
        }
      } catch (error) {
        console.error('[DiscordBridge] Failed to verify bot token:', error);
      }
    }
  }

  /**
   * Called when a new chat session is created
   */
  async onNewSession(session: Session): Promise<void> {
    const url = session.metadata?.url || 'Unknown page';

    const embed = {
      title: 'New chat session',
      color: 0x5865f2, // Discord blurple
      fields: [
        { name: 'Visitor', value: session.visitorId, inline: true },
        { name: 'Page', value: url, inline: false },
      ],
      timestamp: new Date().toISOString(),
    };

    try {
      await this.sendEmbed(embed);
    } catch (error) {
      console.error('[DiscordBridge] Failed to send new session notification:', error);
    }
  }

  /**
   * Called when a visitor sends a message.
   * Returns the Discord message ID for edit/delete sync.
   */
  async onVisitorMessage(
    message: Message,
    session: Session
  ): Promise<BridgeMessageResult> {
    const embed = {
      author: {
        name: session.visitorId,
        icon_url: this.avatarUrl,
      },
      description: message.content,
      color: 0x57f287, // Green
      timestamp: new Date().toISOString(),
    };
    let replyToMessageId: string | undefined;

    if (message.replyTo && this.pocketping?.getStorage().getBridgeMessageIds) {
      const ids = await this.pocketping
        .getStorage()
        .getBridgeMessageIds(message.replyTo);
      if (ids?.discordMessageId) {
        replyToMessageId = ids.discordMessageId;
      }
    }

    try {
      const messageId = await this.sendEmbed(embed, replyToMessageId);
      return { messageId };
    } catch (error) {
      console.error('[DiscordBridge] Failed to send visitor message:', error);
      return {};
    }
  }

  /**
   * Called when an operator sends a message (for cross-bridge sync)
   */
  async onOperatorMessage(
    message: Message,
    _session: Session,
    sourceBridge?: string,
    operatorName?: string
  ): Promise<void> {
    // Don't echo messages that originated from Discord
    if (sourceBridge === 'discord') {
      return;
    }

    const embed = {
      author: {
        name: operatorName || 'Operator',
        icon_url: this.avatarUrl,
      },
      description: message.content,
      color: 0xfee75c, // Yellow
      timestamp: new Date().toISOString(),
    };

    try {
      await this.sendEmbed(embed);
    } catch (error) {
      console.error('[DiscordBridge] Failed to send operator message:', error);
    }
  }

  /**
   * Called when visitor starts/stops typing
   */
  async onTyping(_sessionId: string, isTyping: boolean): Promise<void> {
    if (!isTyping || this.mode !== 'bot' || !this.channelId) return;

    try {
      await fetch(
        `https://discord.com/api/v10/channels/${this.channelId}/typing`,
        {
          method: 'POST',
          headers: { Authorization: `Bot ${this.botToken}` },
        }
      );
    } catch (error) {
      console.error('[DiscordBridge] Failed to send typing indicator:', error);
    }
  }

  /**
   * Called when a visitor edits their message.
   * @returns true if edit succeeded, false otherwise
   */
  async onMessageEdit(
    _messageId: string,
    newContent: string,
    bridgeMessageId: string | number
  ): Promise<boolean> {
    try {
      if (this.mode === 'webhook' && this.webhookUrl) {
        // Webhooks can edit their own messages
        const response = await fetch(
          `${this.webhookUrl}/messages/${bridgeMessageId}`,
          {
            method: 'PATCH',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
              embeds: [
                {
                  description: `${newContent}\n\n*(edited)*`,
                  color: 0x57f287,
                },
              ],
            }),
          }
        );
        return response.ok;
      } else if (this.mode === 'bot' && this.channelId) {
        const response = await fetch(
          `https://discord.com/api/v10/channels/${this.channelId}/messages/${bridgeMessageId}`,
          {
            method: 'PATCH',
            headers: {
              'Content-Type': 'application/json',
              Authorization: `Bot ${this.botToken}`,
            },
            body: JSON.stringify({
              embeds: [
                {
                  description: `${newContent}\n\n*(edited)*`,
                  color: 0x57f287,
                },
              ],
            }),
          }
        );
        return response.ok;
      }
      return false;
    } catch (error) {
      console.error('[DiscordBridge] Failed to edit message:', error);
      return false;
    }
  }

  /**
   * Called when a visitor deletes their message.
   * @returns true if delete succeeded, false otherwise
   */
  async onMessageDelete(
    _messageId: string,
    bridgeMessageId: string | number
  ): Promise<boolean> {
    try {
      if (this.mode === 'webhook' && this.webhookUrl) {
        const response = await fetch(
          `${this.webhookUrl}/messages/${bridgeMessageId}`,
          { method: 'DELETE' }
        );
        return response.ok || response.status === 404;
      } else if (this.mode === 'bot' && this.channelId) {
        const response = await fetch(
          `https://discord.com/api/v10/channels/${this.channelId}/messages/${bridgeMessageId}`,
          {
            method: 'DELETE',
            headers: { Authorization: `Bot ${this.botToken}` },
          }
        );
        return response.ok || response.status === 404;
      }
      return false;
    } catch (error) {
      console.error('[DiscordBridge] Failed to delete message:', error);
      return false;
    }
  }

  /**
   * Called when a custom event is triggered from the widget
   */
  async onCustomEvent(
    event: { name: string; data?: Record<string, unknown> },
    session: Session
  ): Promise<void> {
    const embed = {
      title: `Custom Event: ${event.name}`,
      color: 0xeb459e, // Fuchsia
      fields: [
        { name: 'Visitor', value: session.visitorId, inline: true },
        ...(event.data
          ? [
              {
                name: 'Data',
                value: `\`\`\`json\n${JSON.stringify(event.data, null, 2)}\n\`\`\``,
                inline: false,
              },
            ]
          : []),
      ],
      timestamp: new Date().toISOString(),
    };

    try {
      await this.sendEmbed(embed);
    } catch (error) {
      console.error('[DiscordBridge] Failed to send custom event:', error);
    }
  }

  /**
   * Called when a user identifies themselves via PocketPing.identify()
   */
  async onIdentityUpdate(session: Session): Promise<void> {
    if (!session.identity) return;

    const identity = session.identity;
    const fields = [
      { name: 'User ID', value: identity.id, inline: true },
    ];

    if (identity.name) {
      fields.push({ name: 'Name', value: identity.name, inline: true });
    }
    if (identity.email) {
      fields.push({ name: 'Email', value: identity.email, inline: true });
    }

    const embed = {
      title: 'User Identified',
      color: 0x5865f2,
      fields,
      timestamp: new Date().toISOString(),
    };

    try {
      await this.sendEmbed(embed);
    } catch (error) {
      console.error('[DiscordBridge] Failed to send identity update:', error);
    }
  }

  // ─────────────────────────────────────────────────────────────────
  // Private helper methods
  // ─────────────────────────────────────────────────────────────────

  /**
   * Send an embed to Discord
   */
  private async sendEmbed(
    embed: Record<string, unknown>,
    replyToMessageId?: string
  ): Promise<string | undefined> {
    const body: Record<string, unknown> = {
      embeds: [embed],
    };

    if (this.username) {
      body.username = this.username;
    }
    if (this.avatarUrl) {
      body.avatar_url = this.avatarUrl;
    }

    if (this.mode === 'webhook' && this.webhookUrl) {
      // Use ?wait=true to get the message back
      if (replyToMessageId) {
        body.message_reference = { message_id: replyToMessageId };
      }
      const response = await fetch(`${this.webhookUrl}?wait=true`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Discord webhook error: ${error}`);
      }

      const data = (await response.json()) as DiscordMessageResponse;
      return data.id;
    } else if (this.mode === 'bot' && this.channelId) {
      if (replyToMessageId) {
        body.message_reference = { message_id: replyToMessageId };
      }
      const response = await fetch(
        `https://discord.com/api/v10/channels/${this.channelId}/messages`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: `Bot ${this.botToken}`,
          },
          body: JSON.stringify(body),
        }
      );

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Discord API error: ${error}`);
      }

      const data = (await response.json()) as DiscordMessageResponse;
      return data.id;
    }

    return undefined;
  }
}
